### PR TITLE
REF: share code for set-like ops in DTI/TDI/PI

### PR DESCRIPTION
**Pull request by jbrockmendel**
The DatetimeIndexOpsMixin._concat_same_type this has can probably become the ExtensionIndex._concat_same_type default.  I expect IntervalIndex can be updated to look the same pretty easily, not so sure about CategoricalIndex based on a quick look.

### CODE REVIEW

File: pandas/core/arrays/datetimelike.py [hunk 1]
```diff
@@ -691,12 +691,31 @@ def take(self, indices, allow_fill=False, fill_value=None):
...
+
+        if is_period_dtype(to_concat[0].dtype):
+            new_freq = obj.freq
+        else:
+            # GH 3232: If the concat result is evenly spaced, we can retain the
+            # original frequency
+            new_freq = None
+            to_concat = [x for x in to_concat if len(x)]
+
+            if obj.freq is not None and all(x.freq == obj.freq for x in to_concat):
```
Comment (jreback) [inline pandas/core/arrays/datetimelike.py]: why don't we just infer the freq? seems less complex (or is this *significantly* cheaper?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/arrays/datetimelike.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/core/arrays/datetimelike.py]: This is about 3x faster than infer_freq, it also is logically equivalent to what we're doing now (infer_freq wouldn't be) and about 2x faster.

**Comment by jreback** [comment]
can you merge master here and i'll look again

File: pandas/core/indexes/period.py [hunk 2]
```diff
@@ -697,12 +697,28 @@ def _assert_can_do_setop(self, other):
         if isinstance(other, PeriodIndex) and self.freq != other.freq:
             raise raise_on_incompatible(self, other)
 
-    def intersection(self, other, sort=False):
+    def _setop(self, other, sort, opname: str):
```
Comment (jreback) [inline pandas/core/indexes/period.py]: eventually if you can type / doc-string this

**Review by jreback — APPROVED** [review]

File: pandas/core/arrays/datetimelike.py (hunk 1 above)
Comment (jreback) [reply pandas/core/arrays/datetimelike.py]: k, maybe put this in a function (i would bet we use it elsewhere? maybe in tseries.frequency)

**Comment by jreback** [comment]
k some followup requests

**Pull request merged**